### PR TITLE
Remove every unnecessary any type

### PR DESCRIPTION
**Pull request by Akxe**
I did this mainly to add type to `fieldsMap` that ended up showing that it might return `false`! (In edge case)

### CODE REVIEW

**Comment by Akxe** [comment]
@Mikhus These changes should not have affected tests, not 100% why they did. How do you like this pull request?

File: index.ts [hunk 1]
```diff
@@ -283,18 +283,21 @@ function skipTree(skip: string[]): SkipTree {
...
+        let propTree: SkipTree = tree;
 
         for (let i = 0, s = props.length; i < s; i++) {
             const prop = props[i];
             const all = props[i + 1] === '*';
 
             if (!propTree[prop]) {
-                propTree[prop] = i === s - 1 || all ? true : {};
-                all && i++;
+                if (i === s - 1 || all) {
```
Comment (Mikhus) [inline index.ts:293]: I do not see any reason rewriting this to longer code with no added value.

File: index.ts [hunk 2]
```diff
@@ -374,13 +378,16 @@ function traverse<T extends MapResultKey>(
...
 
         if (nodeSkip !== true) {
-            root[name] = root[name] || (nodes.length ? {} : false);
-            nodes.length && traverse(
-                nodes,
-                root[name],
-                opts,
-                nodeSkip,
-            );
+            const value = root[name] = root[name] || (nodes.length ? {} : false);
```
Comment (Mikhus) [inline index.ts:381]: What is the reason for this change?

File: index.ts [hunk 3]
```diff
@@ -400,16 +407,16 @@ function getBranch(tree: MapResult, path?: string): MapResultKey {
         return tree;
     }
 
-	let branch: MapResultKey = tree;
     for (const fieldName of path.split('.')) {
-        if (!branch[fieldName]) {
+        const branch = tree[fieldName];
```
Comment (Mikhus) [inline index.ts:411]: The same here - what is the reson?

**Review by Mikhus — CHANGES_REQUESTED** [review]
Hi @Akxe
Thank you very much for contribution! I totally agree about general idea to avoid using any types, so far, some of the changes seems unnecessary as for me, please, take a look on my comments below. If we can agree the only change would be about typings I would be very happy to merge this pull request.
All kind regards,
Mike

File: index.ts (hunk 2 above)
Comment (Akxe) [reply index.ts:381]: Typescript does not see the connection between `nodes.length` and `root[name]` type. 
However, if the assignment is extracted to its own variable, typescript can then tell (via if), that `value` is `MapResult` and not `false`, thus make the assignment valid.

It could be also rewritten to: (maybe nicer)
```ts
if (nodes.length) {
  traverse(
    nodes,
    root[name] = {},
    opts,
    nodeSkip,
  );
} else {
   root[name] = false;
}
```

File: index.ts (hunk 1 above)
Comment (Akxe) [reply index.ts:293]: Typescript could not infer the type of `propTree[prop]` (previously at line 297) as `SkipTree` because its definition was somewhat hidden.

I would also argue, that the longer code is more readable, but this is an opinion.

-----

Could be resolved via `propTree = propTree[prop] as SkipTree;`, but I think that it is not necessary, ultimately, it is up to you.

Comment (Akxe) [reply index.ts:293]: Also setting `propTree` to `true` will make the other iteration of `props` do nothing because of `if (!propTree[prop])` condition. Thus this will save some execution time! :)

File: index.ts (hunk 3 above)
Comment (Akxe) [reply index.ts:411]: Same as before. See [TS playground](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAsgQwgJSgZwK4BsQGkpgwC8MAZglmrAD7xKqY4DcwUAHhAPYBOIM40OinTY+JAN4wA2gGsCALhhoQ3AJYA7AOYBdRYmGM8BGAF8WwUhnUBjEKs7qY1hFQCCAChVQoe+iJwANGSqUFgAJgByCAC2PkoqGpoAlIoAbpyqYTDiwDAwqqQwntzeUqQh4VGx2knZuXn8JbAkXlBlFZExUNoseSbA-RZWtvaOzlQAQsXevgaiQeWhnbGKympaKTDpmXV51g7KMABG3Ag2ABbEjaWLlV099QVFJ2fW57U5DdfNx6cXvaYBkA) for an explanation.

File: index.ts (hunk 2 above)
Comment (Akxe) [reply index.ts:381]: I would actually root for the new proposed code as it is more readable.
